rclpy/rclpy/executors.py

```diff
@@ -351,7 +351,8 @@ async def _execute_client(self, client, seq_and_response):
         if header is not None:
             try:
                 sequence = header.request_id.sequence_number
-                future = client._pending_requests[sequence]
+                with client._lock:
+                    future = client._pending_requests[sequence]
```

Comment:
IMO, it might be nice to move some of this logic inside `Client`. For example, the business about the lock and payload (header/response). It moves the Client more towards conforming to the Waitable interface:

```python
async def _execute_client(self, client, seq_and_response):
    header, response = seq_and_response
    if header is None:
        return
    future = client.get_request_future(header)
    future._set_executor(self)
    await client.execute(header, response)
```

Then we can use the lock inside `client.get_request_future()` and `client.execute()`.

Alternatively, we could let the execute method do everything:

```python
async def _execute_client(self, client, seq_and_response):
    await client.execute(self, seq_and_response)
```

This kind of refactor might be left to a follow-up PR as well.

Thoughts?

Reply:
What should the function do if the sequence_number is not in the pending requests? Probably shouldn't let the KeyError exception fall through?

Reply:
> What should the function do if the sequence_number is not in the pending requests? Probably shouldn't let the KeyError exception fall through?

I think that letting `KeyError` fall through is fine.
It should be documented in the docstring.

Executor implementation can continue catching the exception in the same way it's currently being done.

Reply:
Ok, I added the get_pending_request method. (I also took the liberty of moving around the remove_pending_request method so that they would be grouped together nicely).